File: TableView Reorder Objc/ViewController.m
//
//  ViewController.m
//  TableView Reorder Objc
//
//  Created by William Peroche on 20/09/18.
//  Copyright © 2018 William Peroche. All rights reserved.
//

#import "ViewController.h"
#import "AccountCell.h"

#define kAnimationDuration 0.25
#define kVisibleAlpha 1.0F
#define kInvisibleAlpha 0.0F
#define kTranslucentAlpha 0.98F
#define kShadowRadius 5.0F
#define kShadowOpacity 0.5F
#define kContextScale 0.0F

@interface ViewController () <UITableViewDelegate, UITableViewDataSource>

@property (strong, nonatomic) NSMutableArray *itemsArray;
@property (strong, nonatomic) UIView *movingCellSnapshotView;
@property (assign, nonatomic) BOOL movingCellIsAnimating;
@property (assign, nonatomic) BOOL movingCellNeedsToShow;
@property (strong, nonatomic) NSIndexPath *movingCellInitialIndexPath;
@property (strong, nonatomic) IBOutlet UITableView *tableView;


@end

@implementation ViewController



- (void)viewDidLoad {
    [super viewDidLoad];
    self.itemsArray = [[NSMutableArray alloc] initWithObjects:@"1",@"2",@"3",@"4",@"5",@"6",@"7",@"8",@"9",@"10", nil];
    [self setupTableView];
    [self setupGesture];
}


- (void)setupTableView
{
    UINib *nib = [UINib nibWithNibName:kAccountCellIdentifier bundle:nil];
    [self.tableView registerNib:nib forCellReuseIdentifier:kAccountCellIdentifier];
    self.tableView.rowHeight = UITableViewAutomaticDimension;
    self.tableView.estimatedRowHeight = 50.0F;
}

- (void)setupGesture
{
    UILongPressGestureRecognizer *longPress = [[UILongPressGestureRecognizer alloc] initWithTarget:self action:@selector(longPressGestureRecognized:)];
    [self.tableView addGestureRecognizer:longPress];
}


- (void)longPressGestureRecognized:(UILongPressGestureRecognizer *)gestureRecognizer
{
    CGPoint locationInView = [gestureRecognizer locationInView:self.tableView];
    NSIndexPath *indexPath = [self.tableView indexPathForRowAtPoint:locationInView];
    
    if (indexPath == nil)
    {
        [self finalizeReorderOfCell];
        return;
    }
    
    switch (gestureRecognizer.state) {
        case UIGestureRecognizerStateBegan:
            [self createSnapshotOfMovingCellWithIndexPath:indexPath locationInView:locationInView];
            break;

        case UIGestureRecognizerStateChanged:
            [self updateModelWithMovingCellWithIndexPath:indexPath locationInView:locationInView];
            break;

        default:
            [self finalizeReorderOfCell];
            break;
    }
    
}


- (UIView *)snapshotOfCell:(UITableViewCell *)inputView
{
    UIGraphicsBeginImageContextWithOptions(inputView.bounds.size, NO, kContextScale);
    [inputView.layer renderInContext:UIGraphicsGetCurrentContext()];
    UIImage *image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    UIView *snapshotView = [[UIImageView alloc] initWithImage:image];
    [snapshotView.layer setMasksToBounds:NO];
    [snapshotView.layer setShadowOffset:CGSizeMake(-5.0F, 0.0F)];
    [snapshotView.layer setShadowRadius:kShadowRadius];
    [snapshotView.layer setShadowOpacity:kShadowOpacity];
    return snapshotView;
}

- (void)createSnapshotOfMovingCellWithIndexPath:(NSIndexPath *)indexPath locationInView:(CGPoint)locationInView
{
    self.movingCellInitialIndexPath = indexPath;
    UITableViewCell *cell = [self.tableView cellForRowAtIndexPath:indexPath];
    if (cell == nil)
    {
        return;
    }
    
    UIView *snapshotView = [self snapshotOfCell:cell];
    self.movingCellSnapshotView = snapshotView;
    CGPoint center = cell.center;
    snapshotView.center = center;
    snapshotView.alpha = kInvisibleAlpha;
    [self.tableView addSubview:snapshotView];
    
    __weak typeof(self) weakSelf = self;
    
    [UIView animateWithDuration:kAnimationDuration animations:^{
        weakSelf.movingCellIsAnimating = YES;
        snapshotView.center = center;
        snapshotView.alpha = kVisibleAlpha;
        cell.alpha = kInvisibleAlpha;
    } completion:^(BOOL finished) {
        if (finished)
        {
            weakSelf.movingCellIsAnimating = NO;
            if (weakSelf.movingCellNeedsToShow)
            {
                weakSelf.movingCellNeedsToShow = NO;
                [UIView animateWithDuration:kAnimationDuration animations:^{
                    cell.alpha = kVisibleAlpha;
                }];
            }
            else
            {
                cell.hidden = YES;
            }
        }
    }];
    
}

- (void)updateModelWithMovingCellWithIndexPath:(NSIndexPath *)indexPath locationInView:(CGPoint)locationInView
{
    if ((self.movingCellSnapshotView == nil) || (self.movingCellInitialIndexPath == nil))
    {
        return;
    }
    
    CGPoint center = self.movingCellSnapshotView.center;
    center.y = locationInView.y;
    self.movingCellSnapshotView.center = center;
    
    if (indexPath != self.movingCellInitialIndexPath)
    {
        NSString *obj = self.itemsArray[self.movingCellInitialIndexPath.section];
        [self.itemsArray removeObjectAtIndex:self.movingCellInitialIndexPath.section];
        [self.itemsArray insertObject:obj atIndex:indexPath.section];
        [self.tableView moveSection:self.movingCellInitialIndexPath.section toSection:indexPath.section];
        self.movingCellInitialIndexPath = indexPath;
    }
}

- (void)finalizeReorderOfCell
{
    if ((self.movingCellInitialIndexPath == nil) || [self.tableView cellForRowAtIndexPath:self.movingCellInitialIndexPath] == nil || (self.movingCellSnapshotView == nil))
    {
        return;
    }
    
    UITableViewCell *cell = [self.tableView cellForRowAtIndexPath:self.movingCellInitialIndexPath];
    
    if (self.movingCellIsAnimating)
    {
        self.movingCellNeedsToShow = YES;
    }
    else
    {
        cell.hidden = NO;
        cell.alpha = kInvisibleAlpha;
    }
    
    __weak typeof(self) weakSelf = self;
    
    [UIView animateWithDuration:kAnimationDuration animations:^{
        weakSelf.movingCellSnapshotView.center = cell.center;
        weakSelf.movingCellSnapshotView.transform = CGAffineTransformIdentity;
        weakSelf.movingCellSnapshotView.alpha = kInvisibleAlpha;
        cell.alpha = kVisibleAlpha;
    } completion:^(BOOL finished) {
        if (finished)
        {
            weakSelf.movingCellInitialIndexPath = nil;
            [weakSelf.movingCellSnapshotView removeFromSuperview];
            weakSelf.movingCellSnapshotView = nil;
        }
    }];
}


- (NSInteger) numberOfSectionsInTableView:(UITableView *)tableView
{
    return self.itemsArray.count;
}

- (NSInteger) tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return 1;
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    AccountCell *cell = (AccountCell *)[self.tableView dequeueReusableCellWithIdentifier:kAccountCellIdentifier];
    cell.titleLabel.text = self.itemsArray[indexPath.section];
    cell.titleLabel.textColor = [UIColor redColor];
    cell.selectionStyle = UITableViewCellSelectionStyleNone;
    return cell;
}

@end

File: TableView Reorder/ViewController.swift
//
//  ViewController.swift
//  TableView Reorder
//
//  Created by William Peroche on 19/09/18.
//  Copyright © 2018 William Peroche. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    var itemsArray = ["1","2","3","4","5","6","7","8"]
    
    var movingCellSnapshotView: UIView?
    var movingCellIsAnimating = false
    var movingCellNeedsToShow = false
    var movingCellInitialIndexPath: IndexPath?
    
    
    let kAnimationDuration: TimeInterval = 0.25
    let kVisibleAlpha: CGFloat = 1
    let kInvisibleAlpha: CGFloat = 0
    let kTranslucentAlpha: CGFloat = 0.98
    let kTransformScaleXY: CGFloat = 1.0
    @IBOutlet var tableView: UITableView!
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        setupTableView()
        setupGesture()
    }

    func setupGesture() {
        let longpress = UILongPressGestureRecognizer(target: self, action: #selector(ViewController.longPressGestureRecognized(_:)))
        self.tableView.addGestureRecognizer(longpress)
    }
    
    func setupTableView() {
        let nib = UINib(nibName: AccountCell.kAccountCellIdentifier, bundle: nil)
        self.tableView.register(nib, forCellReuseIdentifier: AccountCell.kAccountCellIdentifier)
        self.tableView.rowHeight = UITableViewAutomaticDimension
        self.tableView.estimatedRowHeight = 50.0
    }
    
    @objc func longPressGestureRecognized(_ gestureRecognizer: UIGestureRecognizer) {
        
        let longPress = gestureRecognizer as! UILongPressGestureRecognizer
        let state = longPress.state
        let locationInView = longPress.location(in: self.tableView)
        guard let indexPath = self.tableView.indexPathForRow(at: locationInView) else {
            return
        }
        
        switch state {
        case UIGestureRecognizerState.began:
            self.createSnapshotOfMovingCell(indexPath: indexPath, locationInView: locationInView)
            
        case UIGestureRecognizerState.changed:
            self.updateModelWithMovingCell(indexPath: indexPath, locationInView: locationInView)
            
        default:
            self.finalizeReorderOfCell()
        }
    }
    
    func snapshotOfCell(_ inputView: UITableViewCell) -> UIView {
        
        UIGraphicsBeginImageContextWithOptions(inputView.bounds.size, false, 0.0)
        inputView.layer.render(in: UIGraphicsGetCurrentContext()!)
        let image = UIGraphicsGetImageFromCurrentImageContext()! as UIImage
        UIGraphicsEndImageContext()
        let snapshotView : UIView = UIImageView(image: image)
        snapshotView.layer.masksToBounds = false
        snapshotView.layer.shadowOffset = CGSize(width: -5.0, height: 0.0)
        snapshotView.layer.shadowRadius = 2.0
        snapshotView.layer.shadowOpacity = 0.4
        
        return snapshotView
    }
    
    func createSnapshotOfMovingCell(indexPath: IndexPath, locationInView: CGPoint) {
        self.movingCellInitialIndexPath = indexPath
        guard let cell = self.tableView.cellForRow(at: indexPath) else {
            return
        }
        let snapshotView = snapshotOfCell(cell)
        movingCellSnapshotView = snapshotView
        var center = cell.center
        snapshotView.center = center
        snapshotView.alpha = self.kInvisibleAlpha
        self.tableView.addSubview(snapshotView)
        UIView.animate(withDuration: self.kAnimationDuration, animations: { () -> Void in
            center.y = locationInView.y
            self.movingCellIsAnimating = true
            snapshotView.center = center
            snapshotView.transform = CGAffineTransform(scaleX: self.kTransformScaleXY, y: self.kTransformScaleXY)
            snapshotView.alpha = self.kTranslucentAlpha
            cell.alpha = self.kInvisibleAlpha
        }, completion: { (finished) -> Void in
            if finished {
                self.movingCellIsAnimating = false
                if self.movingCellNeedsToShow {
                    self.movingCellNeedsToShow = false
                    UIView.animate(withDuration: self.kAnimationDuration, animations: { () -> Void in
                        cell.alpha = self.kVisibleAlpha
                    })
                } else {
                    cell.isHidden = true
                }
            }
        })
    }
    
    func updateModelWithMovingCell(indexPath: IndexPath, locationInView: CGPoint) {
        guard let snapshotView = self.movingCellSnapshotView,
            let intialIndexPath = self.movingCellInitialIndexPath else {
                return
        }
        var center = snapshotView.center
        center.y = locationInView.y
        snapshotView.center = center
        if indexPath != intialIndexPath {
            self.itemsArray.insert(self.itemsArray.remove(at: intialIndexPath.section), at: indexPath.section)
            self.tableView.moveSection(intialIndexPath.section, toSection: indexPath.section)
            self.movingCellInitialIndexPath = indexPath
        }
    }
    
    func finalizeReorderOfCell() {
        guard let initialIndexPath = self.movingCellInitialIndexPath,
            let cell = self.tableView.cellForRow(at: initialIndexPath),
            let snapshotView = self.movingCellSnapshotView else {
                return
        }
        
        if self.movingCellIsAnimating {
            self.movingCellNeedsToShow = true
        } else {
            cell.isHidden = false
            cell.alpha = self.kInvisibleAlpha
        }
        UIView.animate(withDuration: self.kAnimationDuration, animations: { () -> Void in
            snapshotView.center = cell.center
            snapshotView.transform = CGAffineTransform.identity
            snapshotView.alpha = self.kInvisibleAlpha
            cell.alpha = self.kVisibleAlpha
        }, completion: { (finished) -> Void in
            if finished {
                self.movingCellInitialIndexPath = nil
                snapshotView.removeFromSuperview()
                self.movingCellSnapshotView = nil
            }
        })
    }
}

extension ViewController: UITableViewDataSource {
    func numberOfSections(in tableView: UITableView) -> Int {
        return itemsArray.count
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: AccountCell.kAccountCellIdentifier) as! AccountCell
        cell.titleLabel.text = itemsArray[indexPath.section]
        cell.titleLabel.textColor = UIColor(red: 0.1, green: 0.6, blue: 0.4, alpha: 1)
        cell.selectionStyle = .none
        cell.backgroundColor = UIColor.red
        return cell
        
    }
    
    func tableView(_ tableView: UITableView, heightForFooterInSection section: Int) -> CGFloat {
        if section == itemsArray.count - 1 {
            return 0
        }
        return 5
    }

    func tableView(_ tableView: UITableView, heightForHeaderInSection section: Int) -> CGFloat {
        if section == 0 {
            return 10
        }
        return 5
    }
}

extension ViewController: UITableViewDelegate {
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        print(itemsArray[indexPath.section])
    }
}
